### PR TITLE
perf: replace deepmerge npm-package with a more performant implementation

### PR DESCRIPTION
**Pull request by Uzlopak**
This PR is replacing the very slow deepmerge package with a special implementation derived from the original code. 
Dont get me wrong: I already proposed to TehShrike two PRs to fix the worst performance bottlenecks two months ago. But no reaction. 

See:

https://github.com/TehShrike/deepmerge/pull/246
https://github.com/TehShrike/deepmerge/pull/247

I actually started it by rewriting the deepmerge code to build a function based on the options, like it is done in this project for JSON-stringificatin. But then I realized, that we need for this project just one simple implementation. So I aborted the rewrite and just implemented the only necessary one.

Here the benchmark. merge is this merge, and deepmerge is the package.

```
merge existing simple keys in target at the roots x 1,006,553 ops/sec ±1.21% (88 runs sampled)
deepmerge existing simple keys in target at the roots x 193,263 ops/sec ±0.79% (89 runs sampled)
merge existing simple keys in target at the roots x 7,392,616 ops/sec ±1.56% (89 runs sampled)
deepmerge existing simple keys in target at the roots x 1,591,627 ops/sec ±1.14% (86 runs sampled)
merge nested objects into target x 5,219,415 ops/sec ±1.09% (87 runs sampled)
deepmerge nested objects into target x 27,331 ops/sec ±1.23% (58 runs sampled)
```

```javascript
var Benchmark = require('benchmark');
var deepmerge = require('deepmerge')
var merge = require('./merge')
var suite = new Benchmark.Suite;

var srcSimple = { key1: 'changed', key2: 'value2' }
var targetSimple = { key1: 'value1', key3: 'value3' }


var srcNested = {
  key1: {
    subkey1: 'subvalue1',
    subkey2: 'subvalue2',
  },
}
var targetNested = {
  key1: 'value1',
  key2: 'value2',
}

var ArraySrc = [{ ...srcSimple }, { ...srcSimple }, { ...srcSimple }, { ...srcSimple }, { ...srcSimple }]
var ArrayTarget = [{ ...targetSimple }, { ...targetSimple }, { ...targetSimple }, { ...targetSimple }, { ...targetSimple }]


// add tests
suite.add('merge existing simple keys in target at the roots', function () {
  var res = merge(ArrayTarget, ArraySrc)
})
// add tests
suite.add('deepmerge existing simple keys in target at the roots', function () {
  var res = deepmerge(ArrayTarget, ArraySrc)
})

suite.add('merge existing simple keys in target at the roots', function () {
  var res = merge(targetSimple, srcSimple)
})
// add tests
suite.add('deepmerge existing simple keys in target at the roots', function () {
  var res = deepmerge(targetSimple, srcSimple)
})
suite.add('merge nested objects into target', function () {
  var res = merge(targetNested, srcNested)
})
suite.add('deepmerge nested objects into target', function () {
  var res = deepmerge(targetNested, srcNested)
})
// add listeners
suite.on('cycle', function (event) {
  console.log(String(event.target));
})
// run async
suite.run({ async: true });
```

### CODE REVIEW

File: merge.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
+'use strict'
+
+// based on https://github.com/TehShrike/deepmerge
```
Comment (mcollina) [inline merge.js:3]: please include the original license of that code

File: test/merge.test.js [hunk 2]
```diff
@@ -0,0 +1,533 @@
+// based on https://github.com/TehShrike/deepmerge/tree/master/test
```
Comment (mcollina) [inline test/merge.test.js:1]: Please include the original license of that code

**Review by mcollina — CHANGES_REQUESTED** [review]
Good work! Only a couple of nits

**Comment by mcollina** [comment]
I think we could use a faster version of that modules in a couple more places.

Maybe we should put it in a module? cc @fastify/core

**Comment by dalisoft** [comment]
@mcollina I am vote for make modules within @fastify organization to separate module and keep this module readable

**Review by Fdawgs — COMMENTED** [review]
Needs rebasing as it's reintroducing chalk, that you just recently removed in another PR 😊

**Comment by Uzlopak** [comment]
If you want to have it in a separate module, then I would recommend to create a new project in the org, like we did with the accept-negotatior ;).

**Comment by Uzlopak** [comment]
Closing in favor of #483

**Pull request closed without merging**